foreman/data_refinery_foreman/foreman/main.py

```diff
@@ -0,0 +1,260 @@
+    instances to determine if jobs are hanging due to a lack of
+    instances. A naive time-based implementation like this could end
+    up retrying every single queued job if there were a long period of
+    spot instance bid price being very high.
+    """
+    minimum_creation_time = timezone.now() - MAX_QUEUE_TIME
+    lost_jobs = DownloaderJob.objects.filter(
+        success=None,
+        retried=False,
+        start_time=None,
+        end_time=None,
+        created_at__lt=minimum_creation_time
+    )
+
+    handle_downloader_jobs(lost_jobs)
+
+
+@retry(stop_max_attempt_number=3)
+@transaction.atomic
+def requeue_processor_job(last_job: ProcessorJob) -> None:
+    """Queues a new downloader job and a Celery task for it.
+
+    The new downloader job will have num_retries one greater than
+    last_job.num_retries.
+    """
+    num_retries = last_job.num_retries + 1
+
+    new_job = ProcessorJob.create_job_and_relationships(num_retries=num_retries,
+                                                        batches=list(last_job.batches.all()),
+                                                        pipeline_applied=last_job.pipeline_applied)
+    processor_task = PROCESSOR_PIPELINE_LOOKUP[last_job.pipeline_applied]
+    app.send_task(processor_task, args=[new_job.id])
+
+    last_job.retried = True
+    last_job.success = False
+    last_job.save()
+
+
+def handle_processor_jobs(jobs: List[ProcessorJob]) -> None:
+    """For each job in jobs, either retry it or log it."""
+    for job in jobs:
+        if job.num_retries < MAX_NUM_RETRIES:
+            requeue_processor_job(job)
+        else:
+            handle_repeated_failure(job)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_failed_processor_jobs() -> None:
+    """Handle processor jobs that were marked as a failure."""
+    failed_jobs = ProcessorJob.objects.filter(success=False, retried=False)
+    handle_processor_jobs(failed_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_hung_processor_jobs() -> None:
+    """Retry processor jobs that were started but never finished."""
+    minimum_start_time = timezone.now() - MAX_PROCESSOR_RUN_TIME
+    hung_jobs = ProcessorJob.objects.filter(
+        success=None,
+        retried=False,
+        end_time=None,
+        start_time__lt=minimum_start_time
+    )
+
+    handle_processor_jobs(hung_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_lost_processor_jobs() -> None:
+    """Retry processor jobs who never even got started for too long."""
+    minimum_creation_time = timezone.now() - MAX_QUEUE_TIME
+    lost_jobs = ProcessorJob.objects.filter(
+        success=None,
+        retried=False,
+        start_time=None,
+        end_time=None,
+        created_at__lt=minimum_creation_time
+    )
+
+    handle_processor_jobs(lost_jobs)
+
+
+def monitor_jobs():
+    """Runs a thread for each job monitoring loop."""
+    threads = []
+    thread = Thread(target=retry_failed_downloader_jobs,
```

Comment:
Since the second argument of `Thread()` is always the function's name, do you think it will be better to create a list of functions first, then use a loop to iterate the loop and get each function's name by `f.func_name`? Something like this:
```
functions = [f1, f2, ...]
threads = []
for f in functions: 
    thread = Thread(target=f, name=f.func_name)
    thread.start()
    threads.append(thread)
```